cmd/agent/app/reporter/client_metrics.go

```diff
@@ -0,0 +1,245 @@
+	params        ClientMetricsReporterParams
+	clientMetrics *clientMetrics
+	shutdown      chan struct{}
+	closed        *atomic.Bool
+
+	// map from client-uuid to *lastReceivedClientStats
+	lastReceivedClientStats sync.Map
+}
+
+// ClientMetricsReporterParams is used as input to WrapWithClientMetrics.
+type ClientMetricsReporterParams struct {
+	Reporter        Reporter        // required
+	Logger          *zap.Logger     // required
+	MetricsFactory  metrics.Factory // required
+	ExpireFrequency time.Duration
+	ExpireTTL       time.Duration
+}
+
+// WrapWithClientMetrics creates ClientMetricsReporter.
+func WrapWithClientMetrics(params ClientMetricsReporterParams) *ClientMetricsReporter {
+	if params.ExpireFrequency == 0 {
+		params.ExpireFrequency = defaultExpireFrequency
+	}
+	if params.ExpireTTL == 0 {
+		params.ExpireTTL = defaultExpireTTL
+	}
+	cm := new(clientMetrics)
+	metrics.MustInit(cm, params.MetricsFactory.Namespace(metrics.NSOptions{Name: "client_stats"}), nil)
+	r := &ClientMetricsReporter{
+		params:        params,
+		clientMetrics: cm,
+		shutdown:      make(chan struct{}),
+		closed:        atomic.NewBool(false),
+	}
+	go r.expireClientMetricsLoop()
+	return r
+}
+
+// EmitZipkinBatch delegates to underlying Reporter.
+func (r *ClientMetricsReporter) EmitZipkinBatch(spans []*zipkincore.Span) error {
+	return r.params.Reporter.EmitZipkinBatch(spans)
+}
+
+// EmitBatch processes client data loss metrics and delegates to the underlying reporter.
+func (r *ClientMetricsReporter) EmitBatch(batch *jaeger.Batch) error {
+	r.updateClientMetrics(batch)
+	return r.params.Reporter.EmitBatch(batch)
+}
+
+// Close stops background gc goroutine for client stats map.
+func (r *ClientMetricsReporter) Close() {
+	if r.closed.CAS(false, true) {
+		close(r.shutdown)
+	}
+}
+
+func (r *ClientMetricsReporter) expireClientMetricsLoop() {
+	ticker := time.NewTicker(r.params.ExpireFrequency)
+	defer ticker.Stop()
+	for {
+		select {
+		case now := <-ticker.C:
+			r.expireClientMetrics(now)
+		case <-r.shutdown:
+			return
+		}
+	}
+}
+
+func (r *ClientMetricsReporter) expireClientMetrics(t time.Time) {
+	var size int64
+	r.lastReceivedClientStats.Range(func(k, v interface{}) bool {
+		stats := v.(*lastReceivedClientStats)
+		stats.lock.Lock()
+		defer stats.lock.Unlock()
+
+		if !stats.lastUpdated.IsZero() && t.Sub(stats.lastUpdated) > r.params.ExpireTTL {
+			r.lastReceivedClientStats.Delete(k)
+			r.params.Logger.Debug("have not heard from a client for a while, freeing stats",
+				zap.Any("client-uuid", k),
+				zap.Time("last-message", stats.lastUpdated),
+			)
+		}
+		size++
+		return true // keep running through all values in the map
+	})
+	r.clientMetrics.ConnectedClients.Update(size)
+}
+
+func (r *ClientMetricsReporter) updateClientMetrics(batch *jaeger.Batch) {
+	clientUUID := clientUUID(batch)
+	if clientUUID == "" {
+		return
+	}
+	if batch.SeqNo == nil {
+		return
+	}
+	entry, found := r.lastReceivedClientStats.Load(clientUUID)
```

Comment:
A client is always expected to generate a new UUID during the bootstrap, right? Those are not in any way stable IDs.

Reply:
Yes, new client generates new ID. We don't need them to be stable across restarts.

Reply:
"We don't *support*" would be more appropriate: a client using a stable ID will send new sequence numbers, which will be ignored by the server until the sequence number is higher than a previous connection.